From 81a1fca366d9f5b49836e42acdb19f3ff0c8173c Mon Sep 17 00:00:00 2001
From: Pierrick Bouvier <pierrick.bouvier@linaro.org>
Date: Tue, 21 Nov 2023 16:13:21 +0400
Subject: [PATCH] [win32] detect llvm-windres (used for windows-arm64)

When adding preprocessor option for llvm-windres (using clang as
parameter), it fails. Thus, do not add this.

It's needed to be able to compile windows-arm64 version, because MSYS2
toolchain is LLVM based (instead of GCC/binutils).
---
 configure.ac          |  4 ++++
 cygwin/GNUmakefile.in | 11 ++++++++---
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/configure.ac b/configure.ac
index 805cbaae18ee2..588fdec3827a6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -512,6 +512,9 @@ AS_CASE(["$target_os"],
 ],
 [hiuxmpp*], [AC_DEFINE(__HIUX_MPP__)])    # by TOYODA Eizi <toyoda@npd.kishou.go.jp>
 
+USE_LLVM_WINDRES=no
+windres_version=`windres --version | grep LLVM`
+test -z "$windres_version" || USE_LLVM_WINDRES=yes
 
 AC_PROG_LN_S
 AC_PROG_MAKE_SET
@@ -4245,6 +4248,7 @@ AC_SUBST(MINIOBJS)
 AC_SUBST(THREAD_MODEL)
 AC_SUBST(COROUTINE_TYPE, ${coroutine_type})
 AC_SUBST(PLATFORM_DIR)
+AC_SUBST(USE_LLVM_WINDRES)
 
 firstmf=`echo $FIRSTMAKEFILE | sed 's/:.*//'`
 firsttmpl=`echo $FIRSTMAKEFILE | sed 's/.*://'`
diff --git a/cygwin/GNUmakefile.in b/cygwin/GNUmakefile.in
index 09298590304b5..192a8cc7112b3 100644
--- a/cygwin/GNUmakefile.in
+++ b/cygwin/GNUmakefile.in
@@ -3,9 +3,14 @@ gnumake = yes
 include Makefile
 
 ENABLE_SHARED=@ENABLE_SHARED@
 DLLWRAP = @DLLWRAP@ --target=@target_os@ --driver-name="$(CC)"
-windres-cpp := $(CPP) -xc
-windres-cpp := --preprocessor=$(firstword $(windres-cpp)) \
-	$(addprefix --preprocessor-arg=,$(wordlist 2,$(words $(windres-cpp)),$(windres-cpp)))
+ifeq (@USE_LLVM_WINDRES@,yes) # USE_LLVM_WINDRES
+	# llvm-windres fails when preprocessor options are added
+	windres-cpp :=
+else
+	windres-cpp := $(CPP) -xc
+	windres-cpp := --preprocessor=$(firstword $(windres-cpp)) \
+		$(addprefix --preprocessor-arg=,$(wordlist 2,$(words $(windres-cpp)),$(windres-cpp)))
+endif
 WINDRES = @WINDRES@ $(windres-cpp) -DRC_INVOKED
 STRIP = @STRIP@
 
